### Prepare(pytorch 0.3.1)

In [31]:
from __future__ import print_function
import torch

import torch
import torch.autograd as autograd
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim

torch.manual_seed(1)

In [32]:
torch.cuda.is_available()

False

In [33]:
torch.__version__

'0.3.1.post3'

### Tensor as numpy

In [34]:
x=torch.eye(5,5)
x


 1  0  0  0  0
 0  1  0  0  0
 0  0  1  0  0
 0  0  0  1  0
 0  0  0  0  1
[torch.FloatTensor of size 5x5]

In [35]:
x=torch.randn(10,10)
y=torch.randn(10,10)

In [36]:
x+y == torch.add(x,y);
res = torch.zeros(x.size())
torch.add(x,y,out=res);
res==x+y
res==x.add(y)


    1     1     1     1     1     1     1     1     1     1
    1     1     1     1     1     1     1     1     1     1
    1     1     1     1     1     1     1     1     1     1
    1     1     1     1     1     1     1     1     1     1
    1     1     1     1     1     1     1     1     1     1
    1     1     1     1     1     1     1     1     1     1
    1     1     1     1     1     1     1     1     1     1
    1     1     1     1     1     1     1     1     1     1
    1     1     1     1     1     1     1     1     1     1
    1     1     1     1     1     1     1     1     1     1
[torch.ByteTensor of size 10x10]

In [37]:
#reshape
x.view(2,-1)



Columns 0 to 9 
 0.6614  0.2669  0.0617  0.6213 -0.4519 -0.1661 -1.5228  0.3817 -1.0276 -0.5631
-0.3968 -0.6571 -1.6428  0.9803 -0.0421 -0.8206  0.3133 -1.1352  0.3773 -0.2824

Columns 10 to 19 
-0.8923 -0.0583 -0.1955 -0.9656  0.4224  0.2673 -0.4212 -0.5107 -1.5727 -0.1232
-2.5667 -1.4303  0.5009  0.5438 -0.4057  1.1341 -1.1115  0.3501 -0.7703 -0.1473

Columns 20 to 29 
 3.5870 -1.8313  1.5987 -1.2770  0.3255 -0.4791  1.3790  2.5286  0.4107 -0.9880
 0.6272  1.0935  0.0939  1.2381 -1.3459  0.5119 -0.6933 -0.1668 -0.9999 -1.6476

Columns 30 to 39 
-0.9081  0.5423  0.1103 -2.2590  0.6067 -0.1383  0.8310 -0.2477 -0.8029  0.2366
 0.8098  0.0554  1.1340 -0.5326  0.6592 -1.5964 -0.3769 -3.1020 -0.0995 -0.7213

Columns 40 to 49 
 0.2857  0.6898 -0.6331  0.8795 -0.6842  0.4533  0.2912 -0.8317 -0.5525  0.6355
 1.2708 -0.0020 -1.0952  0.6016  0.6984 -0.8005  1.5381  1.4673  1.5951 -1.5279
[torch.FloatTensor of size 2x50]

### Variable wrap Tensor

In [45]:
### Variable

v_x = autograd.Variable(x,requires_grad=True);
v_y = autograd.Variable(y,requires_grad=True);
v_z = v_x + v_y

v_z.data


 1.6770  0.0650 -1.2248  1.4444 -1.0620 -1.4621 -2.4662  1.0500  0.1352 -0.8859
 0.9859 -0.6249  0.2061 -1.0809  0.7395  0.8302  0.4450 -0.8635 -1.2244  1.0138
 3.2531 -3.3037  2.3283 -1.4082 -0.3113  0.5638  1.8693  3.5604 -0.1882  0.6135
-1.9815 -0.6750  0.7575 -2.3002  0.4292 -0.6383  1.6983 -0.5209 -1.2637  0.1375
 0.7585  1.6948 -0.9202 -0.2824 -0.6566  1.0185  0.2797 -0.1611 -1.0454  2.1405
-2.7232  0.9598 -2.5454  1.1540  0.0351 -1.7545  0.4047  0.2588 -0.3103 -0.7883
-2.3743 -0.7142 -0.3110 -0.9179 -0.1730  1.3236 -1.3319  0.4991 -0.7603 -0.2716
 2.0969  0.6984 -0.4162  2.3544 -1.9385  1.4208 -1.7725 -0.8419  0.4084 -3.4932
 0.2443 -0.8539  0.6449 -1.1501  1.0428 -1.2112  0.4121 -1.9657 -0.2142 -0.7393
 1.3199  0.4252 -1.9948  1.1329  1.1018  0.6516 -0.8800  0.2767  2.2915 -0.3983
[torch.FloatTensor of size 10x10]

In [48]:
grad_f =v_z.grad_fn
grad_f

#### backwards gradient

In [49]:
s = v_z.sum()
# start calculate using bp
s .backward()

In [53]:
# dx 
v_x.grad

Variable containing:
    1     1     1     1     1     1     1     1     1     1
    1     1     1     1     1     1     1     1     1     1
    1     1     1     1     1     1     1     1     1     1
    1     1     1     1     1     1     1     1     1     1
    1     1     1     1     1     1     1     1     1     1
    1     1     1     1     1     1     1     1     1     1
    1     1     1     1     1     1     1     1     1     1
    1     1     1     1     1     1     1     1     1     1
    1     1     1     1     1     1     1     1     1     1
    1     1     1     1     1     1     1     1     1     1
[torch.FloatTensor of size 10x10]

In [54]:
# dy
v_y.grad

Variable containing:
    1     1     1     1     1     1     1     1     1     1
    1     1     1     1     1     1     1     1     1     1
    1     1     1     1     1     1     1     1     1     1
    1     1     1     1     1     1     1     1     1     1
    1     1     1     1     1     1     1     1     1     1
    1     1     1     1     1     1     1     1     1     1
    1     1     1     1     1     1     1     1     1     1
    1     1     1     1     1     1     1     1     1     1
    1     1     1     1     1     1     1     1     1     1
    1     1     1     1     1     1     1     1     1     1
[torch.FloatTensor of size 10x10]

In [75]:
#dz? None
type(v_z.grad)

NoneType

#### Linear map

In [71]:
lin = nn.Linear(5, 3)  # maps from R^5 to R^3, parameters A, b
# init by random
print(lin.weight)
print(lin.bias)
#data is 2x5.  A maps from 5 to 3... can we map "data" under A?
data = autograd.Variable(torch.randn(2, 5))
print(lin(data))  # yes

Parameter containing:
-0.4077  0.2774  0.4344 -0.1596  0.0661
-0.2304 -0.2861 -0.2265 -0.2131  0.0849
-0.0513  0.0314 -0.3427 -0.2810 -0.3872
[torch.FloatTensor of size 3x5]

Parameter containing:
-0.3214
 0.3745
 0.1662
[torch.FloatTensor of size 3]

Variable containing:
-0.0527  0.6786  0.2385
 0.0305  0.8032  0.4389
[torch.FloatTensor of size 2x3]



### Bag of Words using logSoftmax ex

用pytorch解决一个多分类的经典问题

model ---> forward计算

lossfunciton 反向bp ,更新参数 ,
loss.backword()
optim.step()

再用 model预测新样本 

In [78]:
data = [("me gusta comer en la cafeteria".split(), "SPANISH"),
        ("Give it to me".split(), "ENGLISH"),
        ("No creo que sea una buena idea".split(), "SPANISH"),
        ("No it is not a good idea to get lost at sea".split(), "ENGLISH")]

test_data = [("Yo creo que si".split(), "SPANISH"),
             ("it is lost on me".split(), "ENGLISH")]

# word_to_ix maps each word in the vocab to a unique integer, which will be its
# index into the Bag of words vector
word_to_ix = {}
for sent, _ in data + test_data:
    for word in sent:
        if word not in word_to_ix:
            word_to_ix[word] = len(word_to_ix)
print(word_to_ix)

VOCAB_SIZE = len(word_to_ix)
NUM_LABELS = 2
class BoWClassifier(nn.Module):  # inheriting from nn.Module!

    def __init__(self, num_labels, vocab_size):
        # calls the init function of nn.Module.  Dont get confused by syntax,
        # just always do it in an nn.Module
        super(BoWClassifier, self).__init__()

        # Define the parameters that you will need.  In this case, we need A and b,
        # the parameters of the affine mapping.
        # Torch defines nn.Linear(), which provides the affine map.
        # Make sure you understand why the input dimension is vocab_size
        # and the output is num_labels!
        self.linear = nn.Linear(vocab_size, num_labels)

        # NOTE! The non-linearity log softmax does not have parameters! So we don't need
        # to worry about that here

    def forward(self, bow_vec):
        # Pass the input through the linear layer,
        # then pass that through log_softmax.
        # Many non-linearities and other functions are in torch.nn.functional
        return F.log_softmax(self.linear(bow_vec), dim=1)


def make_bow_vector(sentence, word_to_ix):
    vec = torch.zeros(len(word_to_ix))
    for word in sentence:
        vec[word_to_ix[word]] += 1
    return vec.view(1, -1)


# 'SPAINIS' 'ENGLISH''SPAINIS' 'ENGLISH''SPAINIS' 'ENGLISH'--->'0,1,0,1'
def make_target(label, label_to_ix):
    return torch.LongTensor([label_to_ix[label]])


model = BoWClassifier(NUM_LABELS, VOCAB_SIZE)

# the model knows its parameters.  The first output below is A, the second is b.
# Whenever you assign a component to a class variable in the __init__ function
# of a module, which was done with the line
# self.linear = nn.Linear(...)
# Then through some Python magic from the Pytorch devs, your module
# (in this case, BoWClassifier) will store knowledge of the nn.Linear's parameters
for param in model.parameters():
    print(param)

# To run the model, pass in a BoW vector, but wrapped in an autograd.Variable
sample = data[0]
bow_vector = make_bow_vector(sample[0], word_to_ix)
log_probs = model(autograd.Variable(bow_vector))
print(log_probs)

{'me': 0, 'gusta': 1, 'comer': 2, 'en': 3, 'la': 4, 'cafeteria': 5, 'Give': 6, 'it': 7, 'to': 8, 'No': 9, 'creo': 10, 'que': 11, 'sea': 12, 'una': 13, 'buena': 14, 'idea': 15, 'is': 16, 'not': 17, 'a': 18, 'good': 19, 'get': 20, 'lost': 21, 'at': 22, 'Yo': 23, 'si': 24, 'on': 25}
Parameter containing:

Columns 0 to 9 
-0.0824  0.1702  0.0907 -0.1331  0.0117  0.1273  0.0936 -0.1253  0.1040  0.0421
-0.1856 -0.1363  0.0314  0.0033 -0.0931  0.1358 -0.1468  0.0609 -0.1194  0.1528

Columns 10 to 19 
 0.1947 -0.0907  0.0538  0.1647  0.1341  0.0176  0.1154  0.1297 -0.0116 -0.0773
-0.1928  0.0616  0.0953  0.1728 -0.1946 -0.1080  0.0449 -0.0586 -0.1770 -0.1454

Columns 20 to 25 
 0.1739 -0.0025  0.0485 -0.0256 -0.1427  0.1659
-0.1518  0.1355 -0.0098 -0.1911 -0.1769 -0.0614
[torch.FloatTensor of size 2x26]

Parameter containing:
1.00000e-02 *
  3.3613
 -5.6078
[torch.FloatTensor of size 2]

Variable containing:
-0.4671 -0.9857
[torch.FloatTensor of size 1x2]



In [79]:
label_to_ix = {"SPANISH": 0, "ENGLISH": 1}

In [87]:
# Run on test data before we train, just to see a before-and-after
for instance, label in test_data:
    # 相当于特征化一个样本 
    bow_vec = autograd.Variable(make_bow_vector(instance, word_to_ix))
    #模型跑一次，like predict
    log_probs = model(bow_vec)
    print(log_probs)

# Print the matrix column corresponding to "creo"
print(next(model.parameters())[:, word_to_ix["creo"]])

loss_function = nn.NLLLoss()
optimizer = optim.SGD(model.parameters(), lr=0.1)

# Usually you want to pass over the training data several times.
# 100 is much bigger than on a real data set, but real datasets have more than
# two instances.  Usually, somewhere between 5 and 30 epochs is reasonable.
for epoch in range(100):
    for instance, label in data:
        # Step 1. Remember that Pytorch accumulates gradients.
        # We need to clear them out before each instance
        model.zero_grad()

        # Step 2. Make our BOW vector and also we must wrap the target in a
        # Variable as an integer. For example, if the target is SPANISH, then
        # we wrap the integer 0. The loss function then knows that the 0th
        # element of the log probabilities is the log probability
        # corresponding to SPANISH
        bow_vec = autograd.Variable(make_bow_vector(instance, word_to_ix))
        target = autograd.Variable(make_target(label, label_to_ix))

        # Step 3. Run our forward pass.
        log_probs = model(bow_vec)

        # Step 4. Compute the loss, gradients, and update the parameters by
        # calling optimizer.step()
        loss = loss_function(log_probs, target)
        loss.backward()
        optimizer.step()

for instance, label in test_data:
    bow_vec = autograd.Variable(make_bow_vector(instance, word_to_ix))
    log_probs = model(bow_vec)
    print('for instance', instance)
    probs = torch.exp(log_probs)
    print('instance log_prob:\n',log_probs)
    print('instance prob:\n',probs)

# Index corresponding to Spanish goes up, English goes down!
print(next(model.parameters())[:, word_to_ix["creo"]])


Variable containing:
-0.0731 -2.6527
[torch.FloatTensor of size 1x2]

Variable containing:
-3.5073 -0.0304
[torch.FloatTensor of size 1x2]

Variable containing:
 0.6999
-0.6981
[torch.FloatTensor of size 2]

for instance ['Yo', 'creo', 'que', 'si']
instance log_prob:
 Variable containing:
-0.0657 -2.7551
[torch.FloatTensor of size 1x2]

instance prob:
 Variable containing:
 0.9364  0.0636
[torch.FloatTensor of size 1x2]

for instance ['it', 'is', 'lost', 'on', 'me']
instance log_prob:
 Variable containing:
-3.6621 -0.0260
[torch.FloatTensor of size 1x2]

instance prob:
 Variable containing:
 0.0257  0.9743
[torch.FloatTensor of size 1x2]

Variable containing:
 0.7252
-0.7234
[torch.FloatTensor of size 2]



#### 得到的参数后的解释

对于`creo`这个词:

Variable containing:
 0.7252
-0.7234
[torch.FloatTensor of size 2]

0.72的权重偏向Spanish -0.72权重偏向English 这就是物理意义

### LeNet For Image Classification ex